orttraining/orttraining/python/training/ortmodule/_fallback.py

```diff
@@ -0,0 +1,207 @@
+                                      _FallbackPolicy.FALLBACK_UNSUPPORTED_INPUT.value: {ORTModuleIOError},
+                                      _FallbackPolicy.FALLBACK_UNSUPPORTED_OUTPUT.value: {ORTModuleIOError},
+                                      _FallbackPolicy.FALLBACK_UNSUPPORTED_TORCH_MODEL.value: {ORTModuleTorchModelException},
+                                      _FallbackPolicy.FALLBACK_UNSUPPORTED_ONNX_MODEL.value: {ORTModuleONNXModelException},
+                                      _FallbackPolicy.FALLBACK_BAD_INITIALIZATION.value: {ORTModuleInitException},
+                                      }
+        self._policy = policy
+        self._exception = None
+
+    def handle_exception(self,
+                          exception: Exception,
+                          policy: Optional[_FallbackPolicy] = None) -> None:
+
+        def _set_exception(policy, exception):
+            '''Sets `exception` into `_FallbackManager` based on the specified `policy`
+
+            If the incoming `exception` is handled by the specified `policy`, then `_FallbackManager`
+            will save the exception as context so that ORTModule can learn about a pending fallback
+            and trigger it during model execution.
+
+            Args:
+                policy (_FallbackPolicy or None): Policy to be checked for the incoming `exception`.
+                    if None is specified, all (except _FallbackPolicy.FALLBACK_DISABLE) are implicitly checked
+                exception (ORTModuleFallbackException): Exception that must be handled
+
+            Raises:
+                Exception: when there is no matching `policy` for the incoming `exception`
+            '''
+
+            if policy is not _FallbackPolicy.FALLBACK_DISABLE and \
+                    self._policy.is_set(policy) and \
+                    (policy.value in self._policy_exception_map and type(exception) in self._policy_exception_map[policy.value]):
+
+                if self._log_level <= _logger.LogLevel.WARNING:
+                    warnings.warn(f'Fallback for policy {policy.name} is pending.', UserWarning)
+                self._exception = exception
+
+        if policy is None:
+            for policy in _FallbackPolicy:
+                _set_exception(policy, exception)
+        else:
+            _set_exception(policy, exception)
+
+        if self._exception is None:
+            raise exception
+
+    def is_pending(self) -> bool:
+        '''Returns True when a fallback is pending
+
+        ORTModule must execute fallback to PyTorch engine when a pending fallback is detected
+        '''
+
+        return self._exception is not None
+
+    def fallback(self, model: torch.nn.Module, *inputs, **kwargs):
+        '''Executes user PyTorch `model` using the provided inputs and return the result'''
+
+        # Pending fallbacks are resetted to enforce retries
+        assert self.is_pending()
+        self._exception = None
```

Comment:
What are the senario would benefit from retry? 

Perf will be bad, if we allow retry, and it endup retrying for every training step.

Reply:
If there is a senario that would benefit from retry, we should also limit the number of retry.

Reply:
This is an open-question topic in the design - I am not 100% sure about this. Maybe after some experience period we can revisit

I have conjectured that some scenarios might benefit from retry, assuming ORT is faster than PyTorch. The more we go for ORT, the faster it should be.. Examples could be when the user uses input that is supported by pytorch but not onnx in one of the batches only, such as `int` or a custom type. Or maybe if the user assigns 2 devices for a single model during evaluation, but return to 1 device during training. Or given the input, the export trigger different branches on the model, allowing or causing an issue during export, etc

Reply:
There is a `persist_fallback` flag to control this behavior. We can benchmark and decide the best default

Reply:
Should we make persist_fallback true as default? I thought when there's failure, most likely it will fail again for the same reason. Keeps retrying won't be beneficial in most cases.